### ISSUE TITLE
windows: '~' Home reference bug 

### ISSUE THREAD

**Issue by yuanjs-0108** [PowerShell/PowerShell#20761]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

when i update PowerShell in my windows 11 workstation version to 7.4.0 below is version
$ $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

when i use the `~` to refer my work home, that may conjunct with current path; at old version that method like resolve `~` to $HOME environment variation.
Hope can resolve this problem. thanks

### Expected behavior

```console
ps> more ~\test
1
```

### Actual behavior

```console
# xxxxxx @ SHC5CG33131WZL in ~\AppData\Local\Temp\tmphcnstt.tmp [18:41:39]
$ more ~\test
Cannot access file C:\Users\xxxxxx\AppData\Local\Temp\tmphcnstt.tmp\~\test
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by rhubarb-geek-nz** [comment]
My understanding is the implementation of tilde expansion is Linux/macOS specific.

https://github.com/PowerShell/PowerShell/issues/20754

**Comment by rhubarb-geek-nz** [comment]
Curious. It works for `type` and `Get-Content`, but not `more`.

Is it really using more.com?

```
c:\windows\system32\more.com
```

It looks like you found the issue when using an external command.

I suggest this is not a bug because `C:\WINDOWS\system32\cmd.exe` does not expand tilde so existing native commands are not designed to handle that expansion.

**Comment by yuanjs-0108** [comment]
it actually use more.com in powershell, and it still not work in such like 
code.exe(vscode) --diff 
This kinds of extend command that need path as parameters.
So have u fix it or add this feature?

**Comment by rhubarb-geek-nz** [comment]
> So have u fix it or add this feature?

I suggested

> Work around, if you mean $HOME in PowerShell then use $HOME.

**Comment by 237dmitry** [comment]
> it actually use more.com in powershell

If you about powershell.exe you are confusing a function with an application

```powershell
Get-Command more*

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Function        more
Application     more.com                                           10.0.22... C:\WINDOWS\system32\more.com
```

**Comment by rhubarb-geek-nz** [comment]
PowerShell Core does not have the function

```
PowerShell 7.4.0
PS> Get-Command more*

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Application     more.com                                           10.0.2262… C:\WINDOWS\system32\more.com
```

**Comment by mklement0** [comment]
To summarize:

* On Windows, `~` is only recognized by the [provider](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Providers) cmdlets, such as `Get-Content` and `Get-ChildItem` (there it refers to the home location of the provider underlying the current location; for the `FileSystem` providers, that is `$HOME`, i.e. the current user's home directory (same as `$env:USERPROFILE` on Windows).

* External programs - such as `more.com` or `code` - do _not_ understand `~` and - unlike on _Unix_-like platforms - PowerShell performs _no_ translation of `~` to the value of `$HOME` behind the scenes.
  * See https://github.com/PowerShell/PowerShell/issues/20750#issuecomment-1822843581 for a discussion about possibly changing that.

* In _Windows Powershell_ `more ~\test` happened to work, because there `more` refers to a PowerShell function by that name (run `$function:more` to see its body) that _wraps_ `more.com` and passes the file-path argument to `Get-Content`, which, as stated, _does_  understand `~`.

* Because this function was _removed_ in PowerShell Core, `more ~\test` now passes `~\test` _directly_ to `more.com`, which therefore fails (because `more.com` doesn't understand `~`).
   * Up to v7.3 you could use _tab-completion_ to work around that problem, as tab-completing a `~`-prefixed path would instantly convert `~` to the value of `$HOME` - this was changed in 7.4.0; again, see #20750

**Comment by rhubarb-geek-nz** [comment]
> `~` is not converted if quoted.

The  behaviours are the same on Linux

bash
```
$ /bin/echo ~ '~' "~"
/home/bythesea ~ ~
```

PowerShell
```
PS> /bin/echo ~ '~' "~"
/home/bythesea ~ ~
```

**Comment by tats-u** [comment]
> The behaviours are the same on Linux

This means pwsh for Windows has intentionally disabled this feature.

**Comment by rhubarb-geek-nz** [comment]
> This means pwsh for Windows has intentionally disabled this feature.

Yes, the interpretation of ~ in PowerShell is Linux/macOS only at the moment.

Now tilde does have two existing historical interpretations in Windows, one as a common prefix or extension for temporary files, the second in batch files when getting the directory where the batch file comes from, eg `%~dp0`.

**Comment by tats-u** [comment]
> Now tilde does have two existing historical interpretations in Windows, one as a common prefix or extension for temporary files, 

Unlike Unix, we do not have to expand `~name` because we do not have to care about user profile folders of other users in Windows.

> the second in batch files when getting the directory where the batch file comes from, eg %~dp0.

It is not affected because it does not start with `~`.

**Comment by rhubarb-geek-nz** [comment]
> Unlike Unix, we do not have to expand `~user` because we do not have to care about user profile folders of other users in Windows.

Why is that the case? Windows is a multi-user operating system and servers often have multiple accounts and need to be administered. That is the original purpose behind PowerShell.

**Comment by tats-u** [comment]
I forgot Windows Server. However, it will be difficult to take users whose names contain space into account.

**Comment by mklement0** [comment]
Implementing `~<username>` expansion is the subject of this long-standing feature request:
* #12387

As for usernames with spaces (which are rare, as is `~<username>` use in general, but there's no reason not to _fully_ emulate the tilde expansion of POSIX-compatible shells): The spaces need to be `` ` ``-escaped, e.g., ``~jane` doe``

As for bringing the emulation of tilde expansion to _Windows_ (for native programs only, as on Unix): In addition to the discussion in the previously linked issue, there is a also a PR, though it would initially be just an [experimental feature](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Experimental_Features):

* #20402

**Comment by rhubarb-geek-nz** [comment]
> As for usernames with spaces (which are rare, as is `~<username>` use in general, but there's no reason not to _fully_ emulate the tilde expansion of POSIX-compatible shells): The spaces need to be `` ` ``-escaped, e.g., `` ~jane` doe ``

I would really hope that scripts don't have hard coded usernames in them! They should be parameterised, for example the following works on macOS zsh...

```
% ( FOO=root ; echo ~$FOO )
/var/root
```

And ksh

```
$ ( FOO=root ; echo ~$FOO )
/var/root
```

but not in bash

```
$ ( FOO=root ; echo ~$FOO )
~root
```

in sh

```
$ ( FOO=root ; echo ~$FOO )
~root
```

So for true emulation you need to be exactly clear on all the use cases and at what point in the expansion the ~ occurs, given the above example it looks like it is after the environment variable substitution. But zsh and ksh behaves differently to bash and sh

**Comment by mklement0** [comment]
I agree that `~<username>` isn't terribly useful, @rhubarb-geek-nz, and personally I don't miss it.
But, as stated, there's no reason not to _fully_ emulate the tilde expansion features of POSIX-compatible shells.

Good point re the different behaviors in different shells - I was only aware of the Bash behavior, where only a _verbatim_ username works: This is in line with the [POSIX spec](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_01) (you can infer it from the fact that tilde expansion is prescribed to occur _before_ parameter expansion).

This, and the prevalence of Bash, point to the fact that  `~<username>` is primarily an _interactive_ convenience, such as typing `ls ~j<tab>`, which expand to `ls ~jdoe` and therefore refers to user `jdoe`'s home directory.

Thus, one option is to limit PowerShell to emulating the POSIX-compliant / Bash behavior and allow _verbatim_ usernames only.

To also support _variable-based_ usernames, the logic would have to be (as implied by the usual quoting rules of POSIX-compatible shells), with `$user = 'jdoe'`:

* `~jdoe`, `~$user`, `~"$user"` should expand to `jdoe`'s home dir. (but only - as usual - if either followed by _no_ additional characters or an _unquoted `/`_)
* Any other forms shouldn't  (e.g., `'~jdoe'`, `"~$user"`)

This would make it a bit more useful, because it makes it suitable for use _in scripts_ too.

I suggest continuing this discussion in #12387, if necessary.
